### PR TITLE
feat(components): add form component

### LINKED ISSUE #176: [16/17] feat(components): add form component
## Summary
Adapt `form` — react-hook-form context + accessible field plumbing (label/control/description/message wired with `aria-*` + ids).

## Ship
- Exports: `Form` (FormProvider), `FormField` (Controller), `FormItem`, `FormLabel`, `FormControl`, `FormDescription`, `FormMessage`, `useFormField` (+ Props)
- New deps: `react-hook-form`, `@hookform/resolvers`, `zod`
- Composes: `Label` (#162)

## Key remaps
- `FormMessage` error text `text-destructive` → `nx:text-error-foreground`; `FormDescription` → `nx:text-muted-foreground`
- `FormLabel` invalid → `nx:text-error-foreground`; `FormControl` wires `aria-invalid`/`aria-describedby`/`id` (Input already styles `aria-invalid`)
- `data-slot` per part

## Notes
`useFormField` reads field state + ids via context.

## Acceptance
- [ ] Standard component bar (see #161)
- [ ] Stories: BasicValidation (zod resolver), SubmitSuccess (play), ShowErrors (play), WithDataAttributes (play)
- [ ] base-variants opt-in: no

## Definition of Done

After implementing this component, the audit must report clean before merge:

```bash
yarn workspace @nexus/react audit:storybook-coverage --component form
# exit 0 — no `missing` or `drift` findings
```

Per [`testing-react.md` § Definition of Done](https://github.com/nexuslabs-ai/nexus/blob/main/.claude/rules/testing-react.md#definition-of-done) — **Required** for Phase 1 component PRs. You can also invoke the `storybook-coverage-reviewer` subagent (registered under `.claude/agents/` and discoverable by name) — any prompt like _"audit form story coverage"_ shells out to the script and renders findings with paste-ready snippets.

Part of #161 · Depends on #162 (label)

### PR DESCRIPTION
**Pull request by aishvaryagupta**
## Summary

- Adapt shadcn/ui `form` into a Nexus component — react-hook-form context + accessible field plumbing.
- Exports `Form` (FormProvider), `FormField` (Controller), `FormItem`, `FormLabel`, `FormControl`, `FormDescription`, `FormMessage`, `useFormField`, and the part prop types.
- `FormControl` wires the control's `id` / `aria-invalid` / `aria-describedby`; `FormLabel` + `FormMessage` use `error-foreground`, `FormDescription` uses `muted-foreground`; `data-slot` per part; `FormItem` stacks its parts with `nx:gap-2`.
- `useFormField` uses live guard clauses (hooks-first, then `<FormField>` / `<FormItem>` checks) instead of shadcn's `{}`-default dead guard.
- Stories with play-fns: `Default`, `BasicValidation` (zod resolver), `SubmitSuccess`, `ShowErrors`, `WithDataAttributes`, `AllVariants`. Not opted into base-variants (per the issue).
- **Deps:** `react-hook-form` is an **optional `peerDependency`** (`peerDependenciesMeta.optional`) — kept in `devDependencies` for the workspace build/stories — and added to vite `rollupOptions.external`, so it is **not bundled** into `dist/index.mjs`. Bundling it would key rhf's React context to a second module instance, so a consumer's own `useForm()` wouldn't match the bundled `Controller`/`useFormContext`. `@hookform/resolvers` + `zod` stay story-only `devDependencies`; `zod` is pinned via a **root `resolutions`** entry (with a `//resolutions` rationale note) to dedupe `@hookform/resolvers`' broad range (`^3.25.0 || ^4.0.0`) so `zodResolver(schema)` typechecks.

## GitHub Issue

Closes #176

## Test Plan

- [x] `yarn workspace @nexus/react typecheck`
- [x] `yarn lint` (`eslint . --max-warnings 0`)
- [x] `yarn workspace @nexus/react audit:storybook-coverage --component form` — 0 missing / 0 drift
- [x] `yarn size-limit` — ESM 66.3 kB (< 68 kB limit, no bump needed); CSS 7.46 kB
- [ ] `yarn test:storybook` — Form play-fns (`SubmitSuccess` / `ShowErrors` / `WithDataAttributes`) exercised in CI; the browser suite doesn't run locally in the worktree

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/react/package.json [hunk 1]
```diff
@@ -68,6 +70,7 @@
     "@tabler/icons-react": "^3.36.1",
     "class-variance-authority": "^0.7.1",
     "clsx": "^2.1.1",
+    "react-hook-form": "^7.77.0",
```
Comment (examlly-tech) [inline packages/react/package.json]: **Blocking — bundle rhf, not peer.** vite externalizes only react/react-dom/jsx-runtime, so `react-hook-form` bundles into `index.mjs`. A consumer's own `useForm()` then resolves to a different module instance than the bundled `FormProvider`/`useFormContext`, so context won't match. Make it a `peerDependency`; Radix bundles fine because the consumer never instantiates its context.

File: packages/react/src/components/ui/form.tsx [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+ * its `id`, `aria-invalid`, and `aria-describedby` (pointing at the description
+ * and, when invalid, the message).
+ */
+function FormControl(props: FormControlProps) {
+  const { error, formItemId, formDescriptionId, formMessageId } =
+    useFormField();
+
+  return (
+    <Slot
+      data-slot="form-control"
```
Comment (examlly-tech) [inline packages/react/src/components/ui/form.tsx:198]: **Minor — dead data-slot.** Radix Slot lets the child's props win, so a slotted `<Input>` renders `data-slot="input"`, not `form-control`; this value never survives. Acceptable (matches shadcn) — flagging so the data-slot-contract gap is on record.

File: package.json [hunk 3]
```diff
@@ -91,6 +91,7 @@
     ]
   },
   "resolutions": {
-    "minimatch": "^10.1.2"
+    "minimatch": "^10.1.2",
+    "zod": "^4.4.3"
```
Comment (examlly-tech) [inline package.json:96]: **Minor — global pin for a dev-only dep.** zod is imported only in `Form.stories.tsx` (devDep), yet this adds a repo-wide `resolutions` entry. It does fix the duplicate-zod `zodResolver` type clash — keep it with a noted rationale, or scope the dedupe more narrowly.

**Review by examlly-tech — CHANGES_REQUESTED** [review]
## Principal Architect Review — Verdict: CHANGES REQUIRED

### Issues

| # | Severity | File | Issue | Suggestion |
| --- | --- | --- | --- | --- |
| 1 | ❌ Blocking | `packages/react/package.json` | `react-hook-form` is a runtime `dependency`, so it bundles into `dist/index.mjs` (vite externalizes only react/react-dom/jsx-runtime). A consumer's own `useForm()` then resolves to a different rhf module instance than the bundled `FormProvider`/`useFormContext`/`Controller` → context mismatch / version skew. | Make `react-hook-form` a `peerDependency`, add it to vite `rollupOptions.external`, keep it in `devDependencies` for the workspace build/stories. |
| 2 | ❌ Blocking | `package.json` | PR is CONFLICTING. #270 (input-otp) + #271 (command) merged after the base, both dep-bearing; `origin/main` already raised the ESM size-limit to **68 kB** while this branch still has 62 kB. | Rebase onto `origin/main`, rebuild, re-run `size-limit` against 68 kB. |
| 3 | ⚠️ Minor | `package.json` | Root `resolutions: { zod }` is a repo-global pin for a story-only (dev) dependency that never ships. | Keep it (it resolves the duplicate-zod `zodResolver` type clash) with a noted rationale, or scope the dedupe more narrowly. |
| 4 | ⚠️ Minor | `packages/react/src/components/ui/form.tsx` | `FormControl`'s `data-slot="form-control"` never survives — Radix Slot lets the slotted child's `data-slot` win, so `<Input>` renders `data-slot="input"`. | Acceptable (matches shadcn); flagged so the data-slot-contract gap is on record. |

### Architectural Concerns

**Bundling a consumer-instantiated context library.** Radix deps bundle safely because the consumer never instantiates Radix context — both provider and consumers are the bundled copy. Form inverts this: the consumer's `useForm()` creates the `control`/`formState` that the bundled `Controller`/`useFormContext` must read, and rhf context is keyed by module identity. `react-hook-form` should be a `peerDependency`; the "we bundle all non-react deps" precedent does not transfer to a context lib the consumer also instantiates.

### Links
- **PR:** https://github.com/nexuslabs-ai/nexus/pull/272
- **GitHub Issue:** #176

File: packages/react/src/components/ui/Form.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+  FormField,
+  FormItem,
+  FormLabel,
+  FormMessage,
+} from './form';
+import { Input } from './input';
+
+const signUpSchema = z.object({
+  username: z.string().min(2, 'Username must be at least 2 characters'),
+  email: z.string().email('Enter a valid email address'),
```
Comment (examlly-tech) [inline packages/react/src/components/ui/Form.stories.tsx]: **Minor — deprecated zod API.** In zod 4 `z.string().email()` is deprecated (`@deprecated` in `zod/v4/classic/schemas.d.ts`) in favour of top-level `z.email()`. It still runs and lint isn't type-checked, so non-blocking. Use `email: z.email('Enter a valid email address')`.

File: packages/react/src/components/ui/form.tsx [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+ * useFormField
+ *
+ * Reads the active field's validation state plus the `id`s that link its label,
+ * control, description, and message. Must be called inside both a `FormField`
+ * and a `FormItem`.
+ */
+function useFormField() {
+  const fieldContext = React.useContext(FormFieldContext);
+  const itemContext = React.useContext(FormItemContext);
+  const { getFieldState } = useFormContext();
```
Comment (examlly-tech) [inline packages/react/src/components/ui/form.tsx]: **Minor — guard partly unreachable.** Outside a `Form`, `useFormContext()` returns null, so destructuring `getFieldState` here throws a cryptic native error before the friendly `throw` at line 100. Read it into a var, run the `!fieldContext` / `!formContext` guards first, then destructure.

File: packages/react/src/components/ui/form.tsx [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+ */
+function FormControl(props: FormControlProps) {
+  const { error, formItemId, formDescriptionId, formMessageId } =
+    useFormField();
+
+  return (
+    <Slot
+      data-slot="form-control"
+      id={formItemId}
+      aria-describedby={
```
Comment (examlly-tech) [inline packages/react/src/components/ui/form.tsx:200]: **Minor — dangling aria-describedby.** `aria-describedby` always points at `formDescriptionId` even when no `FormDescription` is rendered, leaving an idref to a non-existent element. AT ignores it (no axe failure), but reference it only when a description exists.

File: packages/react/src/components/ui/form.tsx [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+/**
+ * FormMessage
+ *
+ * Validation message for the field. Renders the field's error message, falling
+ * back to its `children`, and renders nothing when there is neither.
+ */
+function FormMessage({ className, children, ...props }: FormMessageProps) {
+  const { error, formMessageId } = useFormField();
+  const body = error ? String(error?.message ?? '') : children;
```
Comment (examlly-tech) [inline packages/react/src/components/ui/form.tsx]: **Minor — redundant optional chain.** Inside the truthy `error` branch, `error?.message` can be `error.message`: `error ? String(error.message ?? '') : children`.

**Review by examlly-tech — COMMENTED** [review]
## SDE2 Review — Verdict: MINOR CHANGES

### Issues

| # | Severity | File | Issue | Suggestion |
| --- | --- | --- | --- | --- |
| 1 | ⚠️ Minor | `packages/react/src/components/ui/Form.stories.tsx` | `z.string().email()` is deprecated in zod 4 (`@deprecated` in `zod/v4/classic/schemas.d.ts`). Lint is non-type-checked, so it does not fail the gate. | `email: z.email('Enter a valid email address')` |
| 2 | ⚠️ Minor | `packages/react/src/components/ui/form.tsx` | The friendly "within `<FormField>`" guard is partly unreachable: outside a `Form`, `const { getFieldState } = useFormContext()` destructures `null` and throws a cryptic native error before the guard at line 100. | Read `useFormContext()` into a var, run the `!fieldContext` / `!formContext` guards first, then destructure. |
| 3 | ⚠️ Minor | `packages/react/src/components/ui/form.tsx` | `aria-describedby` always references `formDescriptionId` even when no `FormDescription` is rendered, leaving a dangling idref (AT ignores it — no axe failure). | Reference `formDescriptionId` only when a description exists, or document it as mandatory. |
| 4 | ⚠️ Minor | `packages/react/src/components/ui/form.tsx` | `error ? String(error?.message ?? '')` — the `error?.` optional chain is redundant inside the truthy `error` branch. | `error ? String(error.message ?? '') : children` |

### Research Flags

| Technology | Issue | Recommendation |
| --- | --- | --- |
| zod 4.4.3 | `z.string().email()` deprecated → top-level `z.email()` (string formats promoted, more tree-shakable); method form still runs. | Use `z.email(msg)` in the story schema. |

### Links
- **PR:** https://github.com/nexuslabs-ai/nexus/pull/272
- **GitHub Issue:** #176

**Comment by aishvaryagupta** [comment]
Addressed both reviews — all changes are in `69516de`.

### Blocking (Architect review)

- **rhf bundling (#1):** moved `react-hook-form` out of `dependencies` → **optional `peerDependency`** (`peerDependenciesMeta: { "react-hook-form": { optional: true } }`) + kept in `devDependencies`, and added it to vite `rollupOptions.external`. Verified it is no longer inlined in `dist/index.mjs` (single external import). I went **optional** rather than a hard peer because only `Form` needs rhf out of ~20 components — a hard peer would force every `@nexus/react` consumer to install it; with the package's `sideEffects` flag, non-Form consumers tree-shake the external import away.
- **Conflict (#2):** rebased onto `origin/main` (now parents `e885cf7`, includes #270/#271). Adopted the 68 kB ESM limit; `size-limit` measures 66.3 kB.

### Minor

- **zod `resolutions` (Architect #3):** kept the root `resolutions.zod` and added a `//resolutions` rationale note — it dedupes `@hookform/resolvers`' broad peer range so `zodResolver()` typechecks (story-only dep, ships in no bundle).
- **`data-slot="form-control"` (Architect #4):** kept as-is per your note — it's a valid fallback for a slotted child that sets no `data-slot`.
- **deprecated zod email (SDE2 #1):** `z.string().email()` → `z.email()`.
- **guard (SDE2 #2):** applied the var-read cleanup — `useFormContext()` is read into a var and `getFieldState` destructured only after the field/item guards. I did **not** add a `!formContext` "within `<Form>`" guard: I tested it, and outside a provider `useFormState` throws first (`Cannot read properties of null (reading '_formState')`) before any guard can run, and rules-of-hooks forbids guarding ahead of that hook — so the guard would be unreachable dead code. The friendly outside-`<Form>` error isn't achievable here.
- **`aria-describedby` (SDE2 #3):** I prototyped detecting `<FormDescription>` via `React.Children` but reverted it — that only sees direct children, so a wrapped `<FormDescription>` would be dropped from `aria-describedby`, trading the harmless dangling idref (AT ignores it, as you noted) for a real "description exists but isn't announced" regression. Kept the description as standard field anatomy (every composition includes one).
- **redundant optional chain (SDE2 #4):** `error?.message` → `error.message`.
